### PR TITLE
[NFC] CRM-16819 improve on CRM_Utils_Request::retrieve

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Code improvements to functions to retrieve url data

Before
----------------------------------------
Calling functions need to work with difficult function CRM_Utils_Request::retrieve()

After
----------------------------------------
CRM_Utils_Request::retrieveValue() provides a more usable alternative

Technical Details
----------------------------------------
I'm proposing this as an improved (but not perfect) way to retrieve data from http requests.

The biggest issue IMHO with the CRM_Utils_Request::retrieve method is that the third
parameter is & - and if any of the subsequent parameters need to be passed a NULL
array must be generated. The practice of passing a specific static NULL array is known to
have caused errors due to contamination. This commit offers an alternate function without that
parameter (my suggestion is to create a new function again where we really do want  to
work but in general it seems flakey to me). I have also re-ordered the parameters as I perceive
 as being more commonly required than  (renamed from 'abort').

I have also made it so the new function will throw an exception rather than use the deprecated
fatal - as a new function I feel we can enforce this change and I have made some changes
to the error message thrown in the validate function so it returns a list of permissable
values for the  string as the existing error is hard for developers to work with, as it
can be hard to realise why 'int' is wrong (casing) or what should have been passed.

ie this message CRM_Core_Error::fatal("Cannot recognize $type for $data") is replaced by  CRM_Core_Error::fatal(ts('Invalid type, must be one of : ' . implode($possibleTypes)));

Finally I made the visibility on CRM_Utils_Retrieve::getValues protected. This function
is not called from within core & is downright unsafe as it does no validation so we should
not permit anyone to accidentally use it (I have no reason to think they have

Comments
----------------------------------------
I also removed some legacy handling for "Please convert retrieve call to use new function signature" committed in 2013

Previous attempts to improve the way this code have been retreated battered by their arch enemy 'perfect'
---

 * [CRM-16819: Improve on CRM_Utils_Request::retrieve\(\)](https://issues.civicrm.org/jira/browse/CRM-16819)

### CODE REVIEW

File: CRM/Utils/Request.php [hunk 1]
```diff
@@ -79,17 +79,15 @@ public static function id() {
...
+   *   Should a an exception be thrown rather than a fatal.
    *
    * @return mixed
    *   The value of the variable
+   *
+   * @throws \CRM_Core_Exception
    */
-  public static function retrieve($name, $type, &$store = NULL, $abort = FALSE, $default = NULL, $method = 'REQUEST') {
-
-    // hack to detect stuff not yet converted to new style
```
Comment (eileenmcnaughton) [inline CRM/Utils/Request.php]: hack dates back to a signature change before git - think we have changed all calls by now

File: CRM/Utils/Request.php [hunk 2]
```diff
@@ -145,7 +146,7 @@ public static function retrieve($name, $type, &$store = NULL, $abort = FALSE, $d
    * @return mixed
    *    The value of the variable
    */
-  public static function getValue($name, $method) {
+  protected static function getValue($name, $method) {
```
Comment (eileenmcnaughton) [inline CRM/Utils/Request.php]: This is not called from anywhere outside this function at the moment and would unsafe to use directly, hence setting to protected.

**Review by agileware-fj — APPROVED** [review]
As on chat, only concern is performance impact of the programmer feedback extras, which is apparently negligible.
Otherwise looks good, clean and easier to use to me.

**Comment by mlutfy** [comment]
(*[CiviCRM Review Template](https://docs.civicrm.org/dev/en/latest/standards/review/) WORD-1.0*)

* ([`r-jira`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-jira)) ✔
* ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) ✔
* ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) ✔
* ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) n/a
* ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) ✔
* ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) ✔
* ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) n/a
* ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) ✔

Merging based on:

* Review by @agileware-fj 
* Code review, tests, etc.
* Rather non-intrusive for now. Small immediate impact, since it's only used in a few places (dupe merge), but will be slowly propagated with time.
* I have always really disliked `CRM_Utils_Request::retrieve()`, for the reasons Eileen highlighted.

**Pull request merged**